v2pre ts bug, not happy with typed const arrow functions

### What version of Remix are you using?

2.0.0-pre.2

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

Upgrading to v2 release candidate from v1.19.3 is fairly painless on mana.wiki, except for this one weird Typescript error that I don't know what might be the sourced from:

![image](https://github.com/remix-run/remix/assets/84349818/fe88e2ea-6050-4029-8938-fdbc273763f8)

```console
$ remix dev -c "npm run dev:node" --manual

 💿  remix dev

 info  building...
X [ERROR] The character ">" is not valid inside a JSX element

    app/utils/swr-fetchers.ts:25:48:
      25 │ export const swrGqlFetcher = <T>(props: Props) => {
         ╵                                                 ^

  TypeScript's TSX syntax interprets arrow functions with a single generic type parameter as an opening JSX element. If you want it to be interpreted as an arrow function instead, you need to add a trailing comma after the type parameter to 
disambiguate:

    app/utils/swr-fetchers.ts:25:29:
      25 │ export const swrGqlFetcher = <T>(props: Props) => {
         │                              ~~~
         ╵                              <T,>


X [ERROR] Unexpected "const"

    app/utils/swr-fetchers.ts:26:3:
      26 │    const { query, variables } = props;
         ╵    ~~~~~


 info  rebuilding... (~ app\utils\swr-fetchers.ts)
 info  building...
X [ERROR] The character ">" is not valid inside a JSX element

    app/utils/swr-fetchers.ts:26:48:
      26 │ export const swrGqlFetcher = <T>(props: Props) => {
         ╵                                                 ^

  TypeScript's TSX syntax interprets arrow functions with a single generic type parameter as an opening JSX element. If you want it to be interpreted as an arrow function instead, you need to add a trailing comma after the type parameter to 
disambiguate:

    app/utils/swr-fetchers.ts:26:29:
      26 │ export const swrGqlFetcher = <T>(props: Props) => {
         │                              ~~~
         ╵                              <T,>


X [ERROR] Unexpected "const"

    app/utils/swr-fetchers.ts:27:3:
      27 │    const { query, variables } = props;
         ╵    ~~~~~


 info  rebuilding... (~ app\utils\swr-fetchers.ts)
 ```

Updating the .ts files so they don't use arrow functions got the build working, otherwise, great release!

* Mana PR: https://github.com/manawiki/core/pull/207/files

### Expected Behavior

Remix shouldn't have an opinion about how I declare functions in typescript files.

### Actual Behavior

Remix is very insistent on me stop using const arrow functions.

> @pcattori Sorry for insisting but that's not why I see in my 1.19.3 app. Just rename a `.tsx` file to `.ts`:
> 
> ```
> ❯ npx remix build
>  info  building... (NODE_ENV=production)
> ✘ [ERROR] Expected ">" but found "className"
> 
>     app/components/location-selector.ts:22:9:
>       22 │     <div className="flex bg-[#e7b348] h-8 items-center justify-cen...
>          │          ~~~~~~~~~
>          ╵          >
> ```

This would be the behaviour I would expect if there was JSX in a TS file, didn't even know theres a way to go around it

--- a/.changeset/gorgeous-carrots-dance.md
+++ b/.changeset/gorgeous-carrots-dance.md
@@ -1,0 +1,33 @@
+---
+"@remix-run/dev": patch
+---
+
+Do not interpret JSX in .ts files
+
+While JSX is supported in `.js` files for compatibility with existing apps and libraries,
+`.ts` files should not contain JSX. By not interpreting `.ts` files as JSX, `.ts` files
+can contain single-argument type generics without needing a comma to disambiguate from JSX:
+
+```ts
+// this works in .ts files
+
+const id = <T>(x: T) => x;
+//          ^ single-argument type generic
+```
+
+```tsx
+// this doesn't work in .tsx files
+
+const id = <T>(x: T) => x;
+//          ^ is this a JSX element? or a single-argument type generic?
+```
+
+```tsx
+// this works in .tsx files
+
+const id = <T,>(x: T) => x;
+//           ^ comma: this is a generic, not a JSX element
+
+const component = <h1>hello</h1>
+//                   ^ no comma: this is a JSX element
+```

--- a/packages/remix-dev/compiler/js/plugins/hmr.ts
+++ b/packages/remix-dev/compiler/js/plugins/hmr.ts
@@ -139,7 +139,7 @@ declare global {
             sourceCode,
             output: {
               contents: resultCode,
-              loader: "tsx",
+              loader: args.path.endsWith(".ts") ? "ts" : "tsx",
               resolveDir: path.dirname(args.path),
             },
           };

